chainermn/communicators/pure_nccl_communicator.py

```diff
@@ -123,6 +127,12 @@ def _assign_for_allreduce_grad(self, grad_dtype, allreduce_grad_dtype,
 
     def _pack_params_to_buffer(self, params, grad_dtype, allreduce_grad_dtype,
                                n_elems, stream):
+        if self.batched_pack_unpack:
+            if self.params_data is None:
+                self.params_data = _memory_utility.ParamsData(params, 'grad')
```

Comment:
Based on Chainer's design principle define-by-run, the network potentially may change at any iteration. Thus we cannot have any assumption on caching or remembering gradient arrays. Given that this params_data may have N params in iteration X, we can't assume params have N params in iteration X+1.

I'm not sure how much cost we pay at rebuilding params_data here, but there should be a mechanism to detect the grads' change of size, shape or numbers (e.g. compare n_params and all pointers to arrays), or possibly rebuilding params_data every iteration.

Reply:
I understand your concern. But this is not for that purpose, say, sharing gradient array sizes and pointers over iteration. This is just for sharing them between `_pack_params_to_buffer` and `_unpack_params_from_buffer` at the same iteration, so that `self.params_data` is set `None` after `_unpack_params_from_buffer` in `_allreduce_grad_async`. But I think current design is a bit confusing, so will change it to avoid confusion.

Reply:
Sorry! My eyes needed some fix.